softioc/builder.py

```diff
@@ -195,11 +198,15 @@ def _waveform(value, fields):
 
         # Special case for [u]int64: if the initial value comes in as 64 bit
         # integers we cannot represent that, so recast it as [u]int32
+        # Special case for array of strings to correctly identify each element
+        # of the array as a string type.
         if datatype is None:
             if initial_value.dtype == numpy.int64:
                 initial_value = numpy.require(initial_value, numpy.int32)
             elif initial_value.dtype == numpy.uint64:
                 initial_value = numpy.require(initial_value, numpy.uint32)
+            elif initial_value.dtype.char == "S":
```

Comment:
```suggestion
            elif initial_value.dtype.char in ("S", "U"):
```
Might need to catch if people pass List[str] rather than List[bytes] in...

Reply:
I've added this case, as well as a simple test without Unicode characters. I'm not sure what the correct result for Unicode characters is; should it end up with the bytes representation of the character as a "byte string" or as an array of numbers?